### ISSUE TITLE
Bug: sourcing .env with '/' magic

### ISSUE THREAD

**Issue by blaggacao** [casey/just#659]
```console
$ cat baz/.env
BAR="content-of-bar-variable-from-env-file-in-baz-dir"

$ cat baz/justfile
foo:
    echo "$BAR"

$ just baz/foo
echo "$BAR"
sh: BAR: unbound variable
error: Recipe `foo` failed on line 2 with exit code 127

$ cd baz

$ just foo
echo "$BAR"
content-of-bar-variable-from-env-file-in-baz-dir
```

`.env` should be read from the current directory of the justfile's recipe that is invoked.

**Comment by casey** [comment]
Nice find!

This is happening because we use `dotenv::dotenv_iter`, which starts from the processes current working directory, which may be different from the Justfile's directory.

This can be switched by doing the file search ourselves, starting from the Justfile's directory, and using `dotenv::from_path_iter` to load the `.env` file, if we find one. This should be a pretty easy fix.

**Comment by blaggacao** [comment]
If no `.env` there, at any rate, it should _also_ step up the path until it finds one and up to the invoking justfile level.

In the above example:
- first prio: `./baz/.env`
- second prio: `./.env`

This is because `./baz/justfile` might have been independently written with `./baz/.env` in mind. (eg. git submodule)

**Comment by casey** [comment]
> If no `.env` there, at any rate, it should _also_ step up the path until it finds one and up to the invoking justfile level.

Currently, it won't stop if at the level of the invoking justfile. This is to facilitate something like a multi-directory project with multiple justfiles, and a single `.env` file in a parent directory.

So it would check `./baz/.env`, `./.env`, and then `../.env` and `../../.env`, up to the root directory.

**Comment by blaggacao** [comment]
Or wait.

~~Let's don't do too much magic. That would be only for bad surprises...~~


~~Let's settle on:~~

- ~~`./baz/justfile` -> `./baz/.env` (only)~~
- ~~`./justfile` -> `./.env` (only)~~

EDIT:
> Currently, it won't stop if at the level of the invoking justfile.

Since it's already magic, we probably shouldn't remove it and leave it checking up to the root directory. Or if not, that's another issue a part from this one.

**Comment by casey** [comment]
> Since it's already magic, we probably shouldn't remove it and leave it checking up to the root directory. Or if not, that's another issue a part from this one.

Yah, agreed. In retrospect, I wish that when I added dotenv loading, I had not made it load a `.env` automatically, but had instead made it require a directive in the justfile, like:

```
set dotenv := ".env"
```

I would also make it so if you wanted to search recursively up to the root for a `.env` file, you would have to give a directory. So:

```
set dotenv := "."
```

Would search in the current directory for a `.env` file, up to the root. (Or something like that)

I might still do that, although it will be a breaking change, so I've been holding off. (Tracked in https://github.com/casey/just/issues/469)

**Comment by casey** [comment]
Fixed in #661! Will cut a release with the fix shortly.
